app/V1Module/presenters/LoginPresenter.php

```diff
@@ -146,9 +153,43 @@ public function checkRefresh() {
    * @throws ForbiddenRequestException
    */
   public function actionRefresh() {
+    $token = $this->getAccessToken();
+
+    $user = $this->getCurrentUser();
+    $this->sendSuccessResponse([
+      "accessToken" => $this->accessManager->issueToken($user, $token->getScopes(), $token->getExpirationTime()),
+      "user" => $this->userViewFactory->getFullUser($user)
+    ]);
+  }
+
+  public function checkIssueToken() {
+    if (!$this->getAccessToken()->isInScope(TokenScope::MASTER)) {
+      throw new ForbiddenRequestException("Restricted tokens cannot be used to issue new tokens");
+    }
+  }
+
+  /**
+   * Issue a new access token with a restricted set of scopes
+   * @POST
+   * @LoggedIn
+   * @Param(type="post", name="scopes", validation="list", description="A list of requested scopes")
+   * @Param(type="post", required=false, name="expiration", validation="integer", description="How long the token should be valid (in seconds)")
+   * @throws ForbiddenRequestException
+   */
+  public function actionIssueToken() {
+    $request = $this->getRequest();
+    // The scopes are not filtered in any way - the ACL won't allow anything that the user cannot do in a full session
+    $scopes = $request->getPost("scopes");
+
+    if (in_array(TokenScope::MASTER, $scopes)) {
```

Comment:
In my opinion there should be further checks on what scopes can user setup on his/hers issued token. We certainly do not want to issue tokens for changing password, etc...